throw `ToolExit` when `--web-port` is an integer outside the valid TCP port range

andrewkolos opened the pull request:
Fixes #123147.

Interestingly, arguing a non-integer value (`asdf` or `123.45`) does not result in a `ToolExit` or crash. Instead, [we default to null](https://github.com/flutter/flutter/blob/f4c3facfdd06b27dc57507bf2995379b85e796dc/packages/flutter_tools/lib/src/isolated/resident_web_runner.dart#L264) and a [randomly-selected open port is used](https://github.com/flutter/flutter/blob/f4c3facfdd06b27dc57507bf2995379b85e796dc/packages/flutter_tools/lib/src/isolated/devfs_web.dart#L204). However, this behavior seems consistent and intentional across different `*port` options across the tool, so I left it alone.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

andrewkolos committed:
add validation

andrewkolos committed:
add test

andrewkolos committed:
cleanup

christopherfujino commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -504,6 +504,23 @@ class RunCommand extends RunCommandBase {
+      }
+      final int? parsed = int.tryParse(arg);
+      if (parsed == null) {
+        return;
```
can we at least add a printError here that we're ignoring what they provided?

andrewkolos committed:
add error message for non-integer --web-port

andrewkolos committed:
experiment: lift port selection logic

andrewkolos committed:
add --no-pub to new test

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into fix-invalid-tcp-port-crash

christopherfujino reviewed (approved):
LGTM thanks!

andrewkolos committed:
move and fix test

andrewkolos committed:
in error messages, remove reference to "--web-port" CLI option

The error messages now originate in ResidentWebRunner, which is conceptually isolated from the CLI. While this makes the error message slightly use useful for the --web-port case, this makes it less likely for the log messages to diverge from the actual port selection logic.

andrewkolos commented:
@christopherfujino re-requesting review because of minor changes in messages and test code

christopherfujino commented on packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:
```diff
@@ -254,15 +255,38 @@ class ResidentWebRunner extends ResidentRunner {
+Received a non-integer value for port: ${debuggingOptions.port}
+A randomly-chosen available port will be used instead.
+''');
+            return globals.os.findFreePort();
```
nice

christopherfujino commented on packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:
```diff
@@ -254,15 +255,38 @@ class ResidentWebRunner extends ResidentRunner {
 
     try {
       return await asyncGuard(() async {
+        Future<int> getPort() async {
```
Does this need to be a local function, since it's only invoked once? Can we just inline this code?

andrewkolos replied:
It doesn't need to be a (local) function, but I don't think a function needs to be called more than once to justify its existence. In this case, I went with a function to

* provide a clear start and end to port selection logic,
* limit the scope of port selection logic, making it more clear that it doesn't interact with any other code within the `asyncGuard` callback, and
* make it easier for first-time readers of the `run` method by moving the low-level details of port selection out of the way (you can fold/hide `getPort` and still understand what `run` does)

amongst other things.

christopherfujino reviewed (approved):
Nit question about inlining `getPort`, but otherwise LGTM
